src/Storages/VirtualColumnUtils.cpp

```diff
@@ -280,8 +281,16 @@ static const ActionsDAG::Node * splitFilterNodeForAllowedInputs(
             auto & node_copy = additional_nodes.emplace_back(*node);
             node_copy.children.clear();
             for (const auto * child : node->children)
-                if (const auto * child_copy = splitFilterNodeForAllowedInputs(child, allowed_inputs, additional_nodes))
+                if (const auto * child_copy = splitFilterNodeForAllowedInputs(child, allowed_inputs, additional_nodes, strict))
+                {
+                    /// Expression like (now_allowed AND allowed) is not allowed if strict = true. This is important for
+                    /// trivial count optimization, otherwise we can get incorrect results. For example, if the query is
+                    /// SELECT count() FROM table WHERE _partition_id = '0' AND rowNumberInBlock() = 1, we cannot apply
+                    /// trivial count.
+                    if (strict)
+                        return nullptr;
                     node_copy.children.push_back(child_copy);
+                }
```

Comment:
Why we check `strict` when we have valid child? Shouldn;t it be like this:
```suggestion
                if (const auto * child_copy = splitFilterNodeForAllowedInputs(child, allowed_inputs, additional_nodes, strict))
                    node_copy.children.push_back(child_copy);
                /// Expression like (now_allowed AND allowed) is not allowed if strict = true. This is important for
                /// trivial count optimization, otherwise we can get incorrect results. For example, if the query is
                /// SELECT count() FROM table WHERE _partition_id = '0' AND rowNumberInBlock() = 1, we cannot apply
                /// trivial count.
                else if (strict)
                    return nullptr;
```

Reply:
Yes, thanks for fixing!